### PR TITLE
Arithmetic operator sugar

### PR DESCRIPTION
**Pull request by wezm**
Allow use of arithmetic operators `+`, `-`, `*`, `/`, which de-sugar into calls to `*_add`, `*_mul`, etc.

### CODE REVIEW

File: fathom/src/surface/distillation.rs [hunk 1]
```diff
@@ -393,16 +393,86 @@ impl<'interner, 'arena, 'env> Context<'interner, 'arena, 'env> {
...
+                    core::Term::FunApp(
+                        core::Term::Prim(
+                            core::Prim::U8Mul
+                            | core::Prim::U16Mul
+                            | core::Prim::U32Mul
+                            | core::Prim::U64Mul
+                            | core::Prim::S8Mul
+                            | core::Prim::S16Mul
+                            | core::Prim::S32Mul
+                            | core::Prim::S64Mul,
```
Comment (brendanzab) [inline fathom/src/surface/distillation.rs:407]: could maybe have a function from prim to binop

File: fathom/src/surface/elaboration.rs [hunk 2]
```diff
@@ -1886,10 +1886,113 @@ impl<'interner, 'arena, 'error> Context<'interner, 'arena, 'error> {
...
+                );
+                let fun_app = core::Term::FunApp(
+                    self.scope.to_scope(core::Term::FunApp(
+                        self.scope.to_scope(fun),
+                        self.scope.to_scope(lhs_expr),
+                    )),
+                    self.scope.to_scope(rhs_expr),
+                );
+
+                (fun_app, output_type)
```
Comment (brendanzab) [inline fathom/src/surface/elaboration.rs:1914]: Might be cleaner to move this all into some `match_bin_op`, perhaps renaming to `synth_bin_op`

File: fathom/src/surface.rs [hunk 3]
```diff
@@ -85,6 +86,46 @@ pub enum Pattern<Range> {
...
+            BinOp::Div(_) => "/",
+        }
+    }
+
+    fn precedence(&self) -> pretty::Prec {
+        match self {
+            BinOp::Add(_) | BinOp::Sub(_) => pretty::Prec::Add,
+            BinOp::Mul(_) | BinOp::Div(_) => pretty::Prec::Mul,
+        }
+    }
```
Comment (brendanzab) [inline fathom/src/surface.rs:120]: Not sure about this being in this module as it starts to introduce weird cycles between the modules… (perhaps something silly to worry about idk). Rust makes it easy to do that but it gives me a bit of pause. We could either pull `Prec` into here, or move this into `pretty`?

**Review by brendanzab — APPROVED** [review]
Looks good! Thanks for this.

File: fathom/src/surface/pretty.rs [hunk 4]
```diff
@@ -293,6 +295,16 @@ impl<'interner, 'arena> Context<'interner, 'arena> {
...
             ),
+            Term::BinOp(_, lhs, op, rhs) => self.paren(
+                prec > op.precedence(),
+                self.concat([
+                    self.term_prec(Prec::Atomic, lhs),
+                    self.space(),
+                    self.text(op.as_str()),
+                    self.space(),
+                    self.term_prec(Prec::Atomic, rhs),
+                ]),
```
Comment (brendanzab) [inline fathom/src/surface/pretty.rs:306]: Oh, the precedences of the operators need to match the grammar

**Review by brendanzab — CHANGES_REQUESTED** [review]

File: fathom/src/surface/pretty.rs (hunk 4 above)
Comment (wezm) [reply fathom/src/surface/pretty.rs]: I pushed some changes for this but I'm not sure if I did it correctly.

**Review by brendanzab — APPROVED** [review]

**Pull request merged**